File: Stack/C++/225_implement_stack_using_queue.cpp
#include <iostream>
#include <queue>
using namespace std;

class MyStack{
    private:
    queue<int> q1,q2;

    public:
    MyStack(){

    }

    void push(int x){               //每次入栈时，先将q1中数据存放在q2中，入栈元素入q1，再将q2元素入q1。从而实现pop元素时后进先出。
        while (!q1.empty())
        {
            q2.push(q1.front());
            q1.pop();
        }
        q1.push(x);
        while (!q2.empty())
        {
            q1.push(q2.front());
            q2.pop();
        }
    }

    int pop(){
        int temp = q1.front();
        q1.pop();
        return temp;
    }

    int top(){
        return q1.front();
    }

    bool empty(){
        return q1.empty();
    }

};

int main(){
    MyStack mystack;
    mystack.push(1);
    mystack.push(2);  
    cout<<mystack.top()<<endl;  
    cout<<mystack.pop()<<endl;
    cout<<mystack.empty()<<endl;

}

File: Stack/python/225_implement_stack_using_queue.py
class MyStack:
    def __init__(self):
        self.a = []
        self.b = []
    
    def push(self, x: int) -> None:
        while self.a:
            self.b.append(self.a.pop())
        self.a.append(x)
        while self.b:
            self.a.append(self.b.pop())

    def pop(self) -> int:
        return self.a.pop();

    def top(self) -> int:
        return self.a[-1];

    def empty(self) -> bool:
        return len(self.a) == 0;

def main():
    mystack = MyStack()
    mystack.push(1)
    mystack.push(2)
    print(mystack.pop())
    print(mystack.top())
    print(mystack.empty())
    
if __name__ == '__main__':
    main()
